ocaml/runtime/caml/stack.h

```diff
@@ -143,6 +143,10 @@ extern uintnat caml_frame_descriptors_mask;
 #define Hash_retaddr(addr) \
   (((uintnat)(addr) >> 3) & caml_frame_descriptors_mask)
 
+#define Retaddr_frame(d) \
+  ((uintnat)&(d)->retaddr_rel + \
+   (uintnat)(intnat)((d)->retaddr_rel))
+
```

Comment:
If `d->retaddr_rel` contains a negative value, it wraps around (signed & unsigned addition give the same answer, but only unsigned has well-defined behaviour on overflow).

`&(d)->retaddr_rel` is the address of the `retaddr_rel` field of the `frame_descr` pointed to by `d`. The offsets are computed relative to the location in which they are stored. (This needs more parens to clarify, I think).

Reply:
> &(d)->retaddr_rel is the address of the retaddr_rel field of the frame_descr pointed to by d. The offsets are computed relative to the location in which they are stored. (This needs more parens to clarify, I think).

ah, yes, it's fine because `caml_frame_descriptors` hold pointers to the actual `caml_frametable` not copies of the frame table entries.